Two minor source code changes to remove warnings

bceverly opened the pull request:
Fixed linker warning about insecure crypto and fixed compiler warning about unsigned being compared to less than zero.

Both of these are probably safe to have hit all builds if someone wants to do a little testing and remove the #ifdef's.  I didn't want to make that assumption for everyone else however so I played it safe.

hardening commented on client/X11/xf_tsmf.c:
```diff
@@ -108,7 +108,11 @@ int xf_tsmf_xv_video_frame_event(TsmfClientContext* tsmf, TSMF_VIDEO_FRAME_EVENT
 		return -1001;
 
 	/* In case the player is minimized */
+#ifdef __OpenBSD__
```
@bceverly can you explain what is specific to OpenBSD here ?
if the check on event->numVisibleRects is useless (which seems to be the case), let's just remove it for all platforms...

bceverly replied:
I'm totally cool with that (and would prefer it) - I just wasn't
comfortable making the call on behalf of everyone else since I am
relatively new here.

My fear is that this was hiding a bug.  Perhaps it needed to be tested to
be == 0 instead of < 0 so I wanted some more experienced eyes to take a
look at it.


Thanks,
Bryan

On Tue, Apr 28, 2015 at 3:59 PM, Hardening <notifications@github.com> wrote:

> In client/X11/xf_tsmf.c
> <https://github.com/FreeRDP/FreeRDP/pull/2583#discussion_r29282025>:
>
> > @@ -108,7 +108,11 @@ int xf_tsmf_xv_video_frame_event(TsmfClientContext* tsmf, TSMF_VIDEO_FRAME_EVENT
> >  		return -1001;
> >
> >  	/* In case the player is minimized */
> > +#ifdef __OpenBSD__
>
> @bceverly <https://github.com/bceverly> can you explain what is specific
> to OpenBSD here ?
> if the check on event->numVisibleRects is useless (which seems to be the
> case), let's just remove it for all platforms...
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/FreeRDP/FreeRDP/pull/2583/files#r29282025>.
>

akallabeth commented on libfreerdp/core/autodetect.c:
```diff
@@ -166,7 +166,11 @@ BOOL autodetect_send_bandwidth_measure_payload(rdpContext* context, UINT16 paylo
+#ifdef __OpenBSD__
+		Stream_Write_UINT32(s, arc4random());
+#else
 		Stream_Write_UINT32(s, rand());
```
Probably a good idea to use ```CryptGenRandom``` from WinPR and do a proper implementation there.

akallabeth commented on libfreerdp/core/autodetect.c:
```diff
@@ -204,7 +208,11 @@ static BOOL autodetect_send_bandwidth_measure_stop(rdpContext* context, UINT16 p
 			{
+#ifdef __OpenBSD__
+				Stream_Write_UINT32(s, arc4random());
+#else
```
Probably a good idea to use ```CryptGenRandom``` from WinPR and do a proper implementation there.

akallabeth commented on libfreerdp/core/autodetect.c:
```diff
@@ -166,7 +166,11 @@ BOOL autodetect_send_bandwidth_measure_payload(rdpContext* context, UINT16 paylo
 	for (i = 0; i < payloadLength / 4; i++)
 	{
+#ifdef __OpenBSD__
+		Stream_Write_UINT32(s, arc4random());
```
Probably a good idea to use ```CryptGenRandom``` from WinPR and do a proper implementation there.

bceverly commented:
OK.  I implemented the feedback in the following way:

1.  For the xf_tsmf.c code, I made a platform-neutral change to look for zero visible rectangles as I think that is more in the spirit of the code comment about what it is doing.  Please review this one carefully because if zero is a legitimate value, I just introduced a bug.

2.  For the arc4random() change, I implemented the BCryptGenerateRandom() function in winpr for all platforms and then leveraged it in the two places where rand() was previously being called.  I removed the loop in the two spots and let the BCryptGenerateRandom() function fill a buffer that was passed in.

Anything further or should this be merged?

Thanks for all the feedback!  I'm getting way more comfortable with the code.

akallabeth commented on libfreerdp/core/autodetect.c:
```diff
@@ -201,11 +203,11 @@ static BOOL autodetect_send_bandwidth_measure_stop(rdpContext* context, UINT16 p
-			{
-				Stream_Write_UINT32(s, rand());
-			}
+			buffer = malloc(payloadLength);
```
unchecked malloc

akallabeth commented on libfreerdp/core/autodetect.c:
```diff
@@ -163,19 +165,19 @@ BOOL autodetect_send_bandwidth_measure_payload(rdpContext* context, UINT16 paylo
-	{
-		Stream_Write_UINT32(s, rand());
-	}
+	buffer = (UCHAR *)malloc(payloadLength);
```
unchecked malloc

akallabeth commented:
Except my two comments +1

akallabeth commented on winpr/libwinpr/bcrypt/bcrypt.c:
```diff
@@ -68,6 +68,36 @@ NTSTATUS BCryptFinishHash(BCRYPT_HASH_HANDLE hHash, PUCHAR pbOutput, ULONG cbOut
+#ifdef __OpenBSD__
+		random_long = arc4random();
+#else
+		random_long = rand();
```
@bceverly maybe a good idea to also try ```/dev/random``` resp ```/dev/urandom```

bceverly commented:
Good catch @akallabeth!  I have pushed up another commit that addresses this.  Please let me know if you want me to make any other changes.

akallabeth commented on libfreerdp/core/autodetect.c:
```diff
@@ -201,11 +208,16 @@ static BOOL autodetect_send_bandwidth_measure_stop(rdpContext* context, UINT16 p
+			if (NULL == buffer)
 			{
-				Stream_Write_UINT32(s, rand());
+				return FALSE;
```
```Stream_Release(s)```, leaking stream here...

akallabeth commented on libfreerdp/core/autodetect.c:
```diff
@@ -163,19 +165,24 @@ BOOL autodetect_send_bandwidth_measure_payload(rdpContext* context, UINT16 paylo
+	if (NULL == buffer)
 	{
-		Stream_Write_UINT32(s, rand());
+		return FALSE;
```
Leaking stream here

akallabeth commented on libfreerdp/core/autodetect.c:
```diff
@@ -163,19 +165,24 @@ BOOL autodetect_send_bandwidth_measure_payload(rdpContext* context, UINT16 paylo
+	BCryptGenRandom(NULL, buffer, payloadLength, 0L);
+	Stream_Write(s, buffer, payloadLength);
+
 	return rdp_send_message_channel_pdu(context->rdp, s, SEC_AUTODETECT_REQ);
```
When ```rdp_send_message_channel_pdu``` fails, the stream is not released

bceverly commented:
I believe I have fixed the stream leaks.  If someone could please review and let me know if there is anything else that I need to do before this is merged, I'd appreciate it.

:-)

bceverly commented:
Actually now that I thinking about it @akallabeth should I be freeing that buffer I malloc'ed before I return from the function?

akallabeth commented:
@bceverly Yes, propably a good idea ^^

bceverly commented:
OK.  Added the code to clean up the memory buffer @akallabeth .  Are we good to merge?

akallabeth commented:
@bceverly Code looks good now.
It would propably be better to use a stronger random implementation if available. ```getentropy``` on openbsd, ```getrandom``` on linux, ```CryptGenRandom ``` on windows and ```/dev/[u]random``` on other unix like systems come to mind.

akallabeth commented:
@bceverly or for the matter, since freerdp depends on openssl, just use the openssl random engine.

hardening commented:
Yep RAND_bytes() is definitely the one to use as we don't need platform specific tricks

bceverly commented:
OK.  I think I've addressed the points raised by @akallabeth and @hardening with this last commit.  Please take one last look at it and, if we are good, let's merge it.

Thanks again everyone!

bceverly commented:
Hey everyone.  Are we good to merge this one?

akallabeth commented:
Looks ok for me, +1

bceverly commented:
Anyone else?  If not, can we please commit?

awakecoding commented:
@freerdp-bot test.

bmiklautz commented:
@bceverly just verified locally. The latest version of your fixes breaks tests.

bceverly commented:
Bernhard,

Do you think it is the thread ID stuff?  If so, let me back it out and see
if that corrects the issue.


Thanks,
Bryan

On Wed, May 20, 2015 at 10:47 AM, Bernhard Miklautz <
notifications@github.com> wrote:

> @bceverly <https://github.com/bceverly> just verified locally. The latest
> version of your fixes breaks tests.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/FreeRDP/FreeRDP/pull/2583#issuecomment-103912853>.
>

bmiklautz commented:
@bceverly I think so but I haven't had the time to check that in detail. But I'll have a look later on.

bceverly commented:
OK - I pushed another version that puts the code inside of an #ifdef
__OpenBSD__ section.  If you could please re-test I would very much
appreciate it.

Thanks.


Thanks,
Bryan

On Wed, May 20, 2015 at 10:53 AM, Bernhard Miklautz <
notifications@github.com> wrote:

> @bceverly <https://github.com/bceverly> I think so but I haven't had the
> time to check that in detail. But I'll have a look later on.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/FreeRDP/FreeRDP/pull/2583#issuecomment-103916155>.
>

bmiklautz commented:
@bceverly with your last fix the test works again. However I'm fairly sure that the real problem here is the && instead of & when you mask the pthread id.

bceverly commented:
doh!  Thanks.  Wow.  Let me fix it right...


Thanks,
Bryan

On Wed, May 20, 2015 at 12:11 PM, Bernhard Miklautz <
notifications@github.com> wrote:

> @bceverly <https://github.com/bceverly> with your last fix the test works
> again. However I'm fairly sure that the real problem here is the && instead
> of & when you mask the pthread id.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/FreeRDP/FreeRDP/pull/2583#issuecomment-103943101>.
>

bceverly commented:
Just pushed it.  If you could please retest when you get a chance Bernhard
I would appreciate it.

How do you do an embarrassed emoticon again?  :-)


Thanks,
Bryan

On Wed, May 20, 2015 at 12:13 PM, Bryan C. Everly <bryan@bceassociates.com>
wrote:

> doh!  Thanks.  Wow.  Let me fix it right...
>
>
> Thanks,
> Bryan
>
> On Wed, May 20, 2015 at 12:11 PM, Bernhard Miklautz <
> notifications@github.com> wrote:
>
>> @bceverly <https://github.com/bceverly> with your last fix the test
>> works again. However I'm fairly sure that the real problem here is the &&
>> instead of & when you mask the pthread id.
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/FreeRDP/FreeRDP/pull/2583#issuecomment-103943101>.
>>
>
>

bmiklautz commented:
@freerdp-bot test

bmiklautz commented:
+1

bceverly commented:
Thanks!


Thanks,
Bryan

On Thu, May 21, 2015 at 10:18 AM, Marc-André Moreau <
notifications@github.com> wrote:

> Merged #2583 <https://github.com/FreeRDP/FreeRDP/pull/2583>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/FreeRDP/FreeRDP/pull/2583#event-310678816>.
>
